org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeLensHandler.java

```diff
@@ -179,45 +83,75 @@ public void acceptSearchMatch(SearchMatch match) throws CoreException {
-				if (type.isInterface() || Flags.isAbstract(type.getFlags())) {
-					CodeLens lens = getCodeLens(IMPLEMENTATION_TYPE, element, unit);
-					lenses.add(lens);
-				}
-			}
+	private static final String PROVIDER = "provider";
+
+	private static final String CLASS = "class";
+
+	private static final String ID = "id";
+
+	private static Set<CodeLensProviderContainerDescriptor> fgContributedCodeLensProviderContainers;
+
+	public List<CodeLensProvider> getExtendedCodeLensProviders() {
+		Set<CodeLensProviderContainerDescriptor> deps = getCodeLensProviderContainerDescriptors();
+		List<CodeLensProvider> res = new ArrayList<>(deps.size());
+		for (CodeLensProviderContainerDescriptor dep : deps) {
+			CodeLensProviderContainer container = dep.getCodeLensProviderContainer();
+			res.addAll(dep.getProviderIds().stream().map(id -> container.getCodeLensProvider(id, preferenceManager)).collect(Collectors.toSet()));
 		}
+
+		return res;
 	}
 
-	private CodeLens getCodeLens(String type, IJavaElement element, ICompilationUnit unit) throws JavaModelException {
-		CodeLens lens = new CodeLens();
-		ISourceRange r = ((ISourceReference) element).getNameRange();
-		final Range range = JDTUtils.toRange(unit, r.getOffset(), r.getLength());
-		lens.setRange(range);
-		String uri = ResourceUtils.toClientUri(JDTUtils.getFileURI(unit));
-		lens.setData(Arrays.asList(uri, range.getStart(), type));
-		return lens;
+	private static synchronized Set<CodeLensProviderContainerDescriptor> getCodeLensProviderContainerDescriptors() {
+		if (fgContributedCodeLensProviderContainers == null) {
+			IConfigurationElement[] elements = Platform.getExtensionRegistry().getConfigurationElementsFor(EXTENSION_POINT_ID);
+			fgContributedCodeLensProviderContainers = Stream.of(elements).map(e -> new CodeLensProviderContainerDescriptor(e)).collect(Collectors.toSet());
+		}
+		return fgContributedCodeLensProviderContainers;
 	}
 
-	private IJavaSearchScope createSearchScope() throws JavaModelException {
-		IJavaProject[] projects = JavaCore.create(ResourcesPlugin.getWorkspace().getRoot()).getJavaProjects();
-		return SearchEngine.createJavaSearchScope(projects, IJavaSearchScope.SOURCES);
+	private static class CodeLensProviderContainerDescriptor {
+
+		private final IConfigurationElement fConfigurationElement;
+
+		private Set<String> fProviderIds;
+
+		private CodeLensProviderContainer fCodeLensProviderContainerInstance;
+
+		public CodeLensProviderContainerDescriptor(IConfigurationElement element) {
+			fConfigurationElement = element;
+
+			IConfigurationElement[] children = fConfigurationElement.getChildren(PROVIDER);
+			fProviderIds = Stream.of(children).map(c -> c.getAttribute(ID)).collect(Collectors.toSet());
+			fCodeLensProviderContainerInstance = null;
+		}
+
+		public Set<String> getProviderIds() {
+			return fProviderIds;
+		}
+
+		public synchronized CodeLensProviderContainer getCodeLensProviderContainer() {
+			if (fCodeLensProviderContainerInstance == null) {
```

Comment:
What happens if a code lens provider is added when the server is running. This code will require vscode to be restarted for a codelens to become effective. Perhaps we should add some bundle change listeners.

Reply:
@gorkem with the current implem, bundles are only discovered at server startup, I don't think bundle change listeners are necessary, unless we revisit the bundle loading strategy